ocean_lib/ocean/ocean_assets.py

```diff
@@ -322,48 +375,185 @@ def create(
-            # Compress DDO
-            document = lzma.compress(ddo_bytes)
+        # Fetch the asset on chain
+        asset = self._get_aquarius(self._metadata_cache_uri).wait_for_asset(did)
 
-        # Only encryption, not compressed
-        elif encrypt_flag and not compress_flag:
-            flags = bytes([2])
-            # Encrypt DDO
-            encrypt_response = DataServiceProvider.encrypt(
-                objects_to_encrypt=ddo_string,
-                encrypt_endpoint=f"{provider_uri}/api/services/encrypt",
+        return asset
+
+    def update(
+        self,
+        did: str,
+        new_metadata: dict,
+        publisher_wallet: Wallet,
+        delete_erc20_tokens: Optional[List[str]] = [],
+        encrypted_files: Optional[str] = None,
+        new_services: Optional[list] = [],
+        new_credentials: Optional[dict] = None,
+        provider_uri: Optional[str] = None,
+        new_erc20_tokens_data: Optional[List[ErcCreateData]] = [],
+        new_deployed_erc20_tokens: Optional[List[ERC20Token]] = [],
+        encrypt_flag: Optional[bool] = False,
+        compress_flag: Optional[bool] = False,
+    ) -> str:
+        """Update an asset on-chain.
+
+        :param did: The DID of the asset to update.
+        :param new_metadata: dict conforming to the Metadata accepted by Ocean Protocol.
+        :param publisher_wallet: Wallet of the publisher updating this asset.
+        :param delete_erc20_tokens: List of existing ERC20 tokens and related Services to delete.
+        :param encrypted_files: str of the files that need to be encrypted before updating.
+        :param new_services: list of new Service objects to add.
+        :param new_credentials: new_credentials necessary for the asset.
+        :param provider_uri: str URL of service provider. This will be used as base to construct the serviceEndpoint for the `access` (download) service
+        :param new_erc20_tokens_data: list of ERC20CreateData necessary for deploying ERC20 tokens for different services.
+        :param new_deployed_erc20_tokens: list of ERC20 tokens which are already deployed.
+        :param encrypt_flag: bool for encryption of the DDO.
+        :param compress_flag: bool for compression of the DDO.
+        :return: update tx_result
+        """
+
+        old_asset = self._get_aquarius(self._metadata_cache_uri).wait_for_asset(did)
+
+        assert old_asset, "Asset not found."
+
+        if new_deployed_erc20_tokens or new_erc20_tokens_data:
+            assert encrypted_files, "Encrypted_files is required for adding new tokens."
+
+        # Check if datatokens in delete_erc20_tokens are in the old_asset
+        for delete_erc20_tokens_data in delete_erc20_tokens:
+            assert delete_erc20_tokens_data in [
+                dt.get("address") for dt in old_asset.datatokens
+            ], f"{delete_erc20_tokens_data} to delete is not in the old asset datatokens."
+
+        # Metadata sanity check
+        assert isinstance(
```

Comment:
I am not sure about separating the parts for the new asset. i.e. checking metadata, files etc. separately? This is an approach that works, but I think Alex's suggestion would be better. I agree that we offer utillities for creating the asset but I think in the case of update we should just grab the changed asset and update it on chain, using all its new features.

Here is a use case:
we update the trusted algorithms on the Asset object, then call update. Do we have to do the trusted algorithms part all over again and put it in the metadata argument instead? Our object already contains the necessary changes.

The drawback on this is flexibility:
- for create, we offer the option to auto-deploy some tokens, which will be limited on update. But I would treat that as a "nothing changed" (same tokens as before).
- in case tokens HAVE changed, we check their addresses and do not offer autodeploy options. Ok, maybe we can work in some ErcCreateData, but after we get the basic flow working.

Bottom line:
We shouldn't update the asset piecemeal, we should use the Asset object.

This is my personal opinion. I will tag @trentmc @alexcos20 for a second opinion though.

Reply:
Sounds good to me, 
I am going to refactor the code to pass directly as a parameter to update the updated asset and thus make the update easier (remove for now the token deployment with erc20 data among other things).